Detect service error

antstorm opened the pull request:
Hi,

I've stumbled across a case when `postgresql` would report being started successfully, but won't have a running process. It turned out that `launchctl` wasn't receiving the error code on `load`, but `list` revealed a non-zero exit code.

With this change in place it will mark services that aren't actually running, although they are loaded.

MikeMcQuaid commented on cmd/brew-services.rb:
```diff
@@ -155,9 +160,18 @@ def list
                   "Name", "Status", "User", "Plist")
       formulae.each do |formula|
+        status = case formula[:status]
+                 when :started
```
Indent `when :started` two-chars after `status`. Sorry for pedantry, a pet hate 😉

antstorm replied:
Happy to do that if that's what you prefer, I'm just following the community style guide here (https://github.com/bbatsov/ruby-style-guide#indent-conditional-assignment)

MikeMcQuaid replied:
Yeh, I normally agree with that but this is one of the few cases I hate; it feels like it's floating in mid-air 😆. Thanks!

antstorm replied:
No worries, mate, pushed that change :)

MikeMcQuaid commented on cmd/brew-services.rb:
```diff
@@ -155,9 +160,18 @@ def list
+                 when :stopped
+                   "stopped"
+                 when :error
+                   "#{Tty.red}error  #{Tty.reset}"
```
Why the two spaces here? For alignment?

antstorm replied:
Yes, exactly! It felt like a proper alignment of the output is a bit out of scope for this PR :)

MikeMcQuaid replied:
Yeh, definitely. Good to explain.

MikeMcQuaid commented on cmd/brew-services.rb:
```diff
@@ -126,21 +126,26 @@ def list
         end
 
+        # Check the exit code of the service, might indicate an error
+        if formula[:status] == :started && service.exit_code.nonzero?
```
`service.exit_code` can be `nil` so that probably needs guarded.

MikeMcQuaid commented on cmd/brew-services.rb:
```diff
@@ -369,8 +383,11 @@ def started_as
+  end
+
+  def exit_code
+    return $2.to_i if status =~ status_regexp
```
An edge-case I noticed with local testing: if I `brew services list` then any services running as `root` will show up as `error` because they aren't listed in the non-root `brew services list` output. Not sure how best to handle that, though.

antstorm commented:
@MikeMcQuaid I've changed the solution a bit to support multiple users. So now it will report service status as "unknown" if we're unable to detect it's status via `launchctl list`. I've also added `pid` check, which is also an indicator that there was an error.

Here's an example running as regular user:

![](https://www.dropbox.com/s/spbwmt6z2oyemfc/Screenshot%202016-12-13%2022.15.49.png?dl=1)

and as root:

![](https://www.dropbox.com/s/75307a5opksrnsr/Screenshot%202016-12-13%2022.16.26.png?dl=1)

P.S. Error next to `mysql` is due to a missing `pid`, which is expected.

MikeMcQuaid commented:
@antstorm Great work. Although the `unknown` is technically correct I wonder if we can use less scary language. Thoughts on just using `started` if it was started as a different user? Alternatively, making it something like `launched` or something?

antstorm commented:
@MikeMcQuaid sure, we can use started, but mark it yellow. Launched also makes sense, but might be a bit confusing since it's a synonym to started. Another option is `started`/`running`.

MikeMcQuaid commented:
I think communicating purely with colours isn't great for usability. Agreed about "launched" problems. Let's maybe just keep it as `started` for now; you're adding detection for a failure case and some people parse the output of this command so I'd probably rather have false-positives than false-negatives here as it still provides some improvement while not requiring anything that parses the output to be changed.

antstorm commented:
@MikeMcQuaid so keep things as they were when showing status of services started by other users?

MikeMcQuaid commented:
Yeh but just for when it would have been "unknown". Basically, rather than showing "unknown" we just fall back to started/stopped like the behaviour before this PR. Progressive enhancement rather than degradation, essentially.

antstorm commented:
@MikeMcQuaid pushed up that change

MikeMcQuaid commented:
Thanks for your truly fantastic contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

antstorm commented:
@MikeMcQuaid thanks, happy to help! :)
